tests/unittests/sources/test_openstack.py

```diff
@@ -603,6 +603,46 @@ def fake_dmi_read(dmi_key):
             "Expected detect_openstack == True on SAP CCloud VM",
         )
 
+    @test_helpers.mock.patch(MOCK_PATH + "dmi.read_dmi_data")
+    def test_detect_openstack_huaweicloud_chassis_asset_tag(
+        self, m_dmi, m_is_x86
+    ):
+        """Return True on OpenStack reporting Huawei Cloud VM asset-tag."""
+        m_is_x86.return_value = True
+
+        def fake_asset_tag_dmi_read(dmi_key):
+            if dmi_key == "system-product-name":
+                return "c7.large.2"  # No match
+            if dmi_key == "chassis-asset-tag":
+                return "HUAWEICLOUD"
+            assert False, "Unexpected dmi read of %s" % dmi_key
+
+        m_dmi.side_effect = fake_asset_tag_dmi_read
+        self.assertTrue(
+            ds.detect_openstack(),
+            "Expected detect_openstack == True on Huawei Cloud VM",
+        )
+
+    @test_helpers.mock.patch(MOCK_PATH + "dmi.read_dmi_data")
+    def test_detect_openstack_huaweicloud_product_name(
+        self, m_dmi, m_is_x86
+    ):
+        """Return True on OpenStack reporting Huawei Cloud VM product name."""
+        m_is_x86.return_value = True
+
+        def fake_product_name_dmi_read(dmi_key):
+            if dmi_key == "system-product-name":
+                return "OpenStack Nova"
+            if dmi_key == "chassis-asset-tag":
+                return "HUAWEICLOUD"
+            assert False, "Unexpected dmi read of %s" % dmi_key
+
+        m_dmi.side_effect = fake_product_name_dmi_read
+        self.assertTrue(
+            ds.detect_openstack(),
+            "Expected detect_openstack == True on Huawei Cloud VM",
+        )
+
```

Comment:
I'm not sure we this unit test here, as the check for `chassis-asset-tag == HWUAWEICLOUD` will be positive before the check for system-product-name  == "OpenStack Nova" in `detect_openstack`. We already have a `test_detect_openstack_intel_product_name_compute` test that exercises `product == OpenStack Nova` anyway.

Reply:
Originally, considering that Huawei cloud has two dmi_key, I wrote test cases respectively.
Agree with your opinion, this test case has been deleted.

Reply:
Thanks